mart/attack/composer.py

```diff
@@ -61,3 +63,63 @@ def compose(self, perturbation, *, input, target):
         mask = mask.to(input)
 
         return input * (1 - mask) + perturbation * mask
+
+
+class ModalityComposer(Composer):
+    """A modality-aware composer.
+
+    Example usage: `ModalityComposer(rgb=Overlay(), depth=Additive())`. Note that
+    `ModalityComposer(default=Additive())` is equivalent with `Additive()`.
+    """
+
+    def __init__(self, **modality_composers):
+        self.modality_composers = modality_composers
+
+    def __call__(
+        self,
+        perturbation: torch.Tensor | tuple,
+        *,
+        input: torch.Tensor | tuple,
+        target: torch.Tensor | dict[str, Any] | tuple,
+        **kwargs,
+    ) -> torch.Tensor | tuple:
+        # Bypass batch-aware in Composer.__call__(), because we have the recursive self.compose().
+        input_adv = self.compose(perturbation, input=input, target=target)
```

Comment:
I would expect, at a minimum, that these arguments (`perturbation`, `input`, and `target`) would have the same types as in `compose` since you're directly passing them to `compose`?

My suspicion is that some functionality probably needs to move to `Composer.__call__` while `dict` functionality should be in `ModalityComposer.__call__`?

Reply:
I feel like it will be less confusing if we follow the implementation in `Enforcer`.